### PR TITLE
Add Entry Point chapter

### PR DESCRIPTION
**Pull request by sfricke-samsung**

### CODE REVIEW

File: chapters/entry_execution.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    foo(3.0);
+}
+```
+
+which gets tranlated to roughly the following in SPIR-V
+
+```swift
+       OpEntryPoint Vertex %v_main "vertex_main" %vert_out
+       OpEntryPoint Fragment %f_main "fragment_main"
+       OpExecutionMode %vertex_main RoundingModeRTE 32
```
Comment (sfricke-samsung) [inline chapters/entry_execution.md:91]:
```suggestion
       OpExecutionMode %v_main RoundingModeRTE 32
```

File: chapters/entry_execution.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+       OpExecutionMode %fragment_main RoundingModeRTZ 32
```
Comment (sfricke-samsung) [inline chapters/entry_execution.md:92]:
```suggestion
       OpExecutionMode %f_main RoundingModeRTZ 32
```

File: chapters/entry_execution.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Entry Point, Execution Model, and Execution Mode
+
+The start of every SPIR-V module will contain the `OpEntryPoint` and `OpExecutionMode`/`OpExecutionModeId` instructions. This chapter is to take a more indepth look into them and edge cases tool developrs will need to watch out for.
```
Comment (dneto0) [inline chapters/entry_execution.md:3]: Suggest rewording  "This chapter takes a more ..."

File: chapters/entry_execution.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+But the following will fail:
+
+```swift
+OpEntryPoint GLCompute %main_a "name_c"
+OpEntryPoint GLCompute %main_b "name_c"
+
+// error: 2 Entry points cannot share the same name and ExecutionMode.
+```
+
+This means any tool will have to be mindful that an **entry point is not unique to single value**. Tool consuming SPIR-V must be aware that two entry points can share the same `Execution Model` or `Name` operand of multiple `OpEntryPoint`.
```
Comment (dneto0) [inline chapters/entry_execution.md:38]: I'm not clear on what the emphasized phrase is supposed to mean exactly, because I'm not sure what "value" refers to.

Do you mean

"This means a simple one-to-one relationship is not guaranteed between:
- entry points and their associated functions
- an entry point name and its associated functions
"

?

For wording on the second sentence, I suggest changing:

> Tool consuming SPIR-V must be aware that two entry points can share the same `Execution Model` or `Name` operand of multiple `OpEntryPoint`.

to 

> Tools consuming SPIR-V must be aware that two entry points can use the same value for the  `Execution Model` or `Name` operands on their `OpEntryPoint` instructions, but not both at the same time.

File: chapters/entry_execution.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+# Execution Mode
+
+The `OpExecutionMode` and `OpExecutionModeId` are simple as they point to an `Entry Point` ID and set the `Execution Mode` for it. The important information is to understand the difference between the two instructions.
```
Comment (dneto0) [inline chapters/entry_execution.md:42]: Suggest removing the "simple" phrasing.   
I recommend:

> The `OpExecutionMode` and `OpExecutionModeId` instructions reference an `Entry Point` ID to set the `Execution Mode` for that entry point.

File: chapters/entry_execution.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+> A notable difference in the spec is that `OpExecutionModeId` was added in SPIR-V 1.2
```
Comment (dneto0) [inline chapters/entry_execution.md:44]: suggest "update" instead of "difference"?

Comment (dneto0) [reply chapters/entry_execution.md:44]: End the note with a period.

File: chapters/entry_execution.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+> Also worth noting that as of now, there are no possible uses of `OpExecutionModeId` for Shader capability-based client APIs such as Vulkan
```
Comment (dneto0) [inline chapters/entry_execution.md:46]: End the note with a period.

File: chapters/entry_execution.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+The only difference between the two instructions is how it handles possible `Extra Operands`.
+
+`OpExecutionMode` uses `Literal`
```
Comment (dneto0) [inline chapters/entry_execution.md:50]: In general, when you have a proper noun that is jargon, you should also use the generic word for it as well:

Instead of:

> `OpExecutionMode` uses `Literal`

do this

> The `OpExecutionMode` instruction uses a `Literal` operand. 

Then introduce the example with:

> For example:

File: chapters/entry_execution.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+OpExecutionModeId %main LocalSizeId %x %y %z
+```
+
+This is noted in the [Execution Mode part of the spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Execution_Mode)
+
+![entry_execution_mode_chart.png](../images/entry_execution_mode_chart.png)
+
+# Instructions with multiple execution modes
+
+It is not obvious at first glance, but instructions in a function block can be ran with multiple execution modes. Imagine the following pseudo code:
```
Comment (dneto0) [inline chapters/entry_execution.md:68]: Suggest changing

> can be ran with multiple execution modes

to

>  can be invoked as part of entry points that have different execution modes.

File: chapters/entry_execution.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+```glsl
+float foo(float bar) {
+    return bar / 2.0;
+}
+
+void vertex_main() {
+    // set RoundingModeRTE
```
Comment (dneto0) [inline chapters/entry_execution.md:76]: You should have a note saying this is psuedo-GLSL because GLSL compilers don't emit more than one entry point per SPIR-V module.

Also I don't think there is a GLSL extension for float-controls.  I didn't find any on the [GLSL registry](https://github.com/KhronosGroup/GLSL), and all the Vulkan conformance tests for float controls used SPIR-V assembly.

**Review by dneto0 — COMMENTED** [review]
Thanks for writing this.  It's a subtle topic.

I've got mostly editorial suggestions.

File: chapters/entry_execution.md (hunk 4 above)
Comment (sfricke-samsung) [reply chapters/entry_execution.md:38]: agree, `value` is not a good word choice here, changed the emphasized part to 

> entry point is not unique to single Execution Model and Name
